### PR TITLE
Support Intel RDT

### PR DESCRIPTION
**Pull request by marquiz**
OCI runtime-spec has supported Intel RDT (resctrl pseudo-filesystem) for a while already. This RFC PR adds two independent parts of RDT support to CRI-O.

The first one is integration with github.com/intel/goresctrl which enables flexible class-based configuration mechanism. The concept of the integration is partly inspired by seccomp et. al. CRI-O config only specifies a path to an external configuration file, and the file structure and logic of applying the configuration is owned by the library. This integration will give users an easy option for configuring RDT in tandem with the container runtime.

The second part adds support for container annotation for controlling the RDT class/CLOS of containers. The idea is that Kubelet can utilize this annotation before the CRI API properly supports RDT. There's also a patch that adds respective Pod annotation for controlling the RDT class (per-pod and per-container) – this is for testing/demonstration purposes of K8s integration.

My take is that the next steps – e.g. how to split this PR, how to proceed with CRI API etc – depend on how this PR is received.

#### What type of PR is this?

/kind feature

> /kind api-change
> /kind bug
> /kind ci
> /kind cleanup
> /kind dependency-change
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind flake
> /kind other

#### What this PR does / why we need it:

The OCI runtime spec has had support for RDT for a long time, already. This PR opens up the path to CRI support. Integration with `goresctrl` is a light and simple extension point for better user experience.

#### Special notes for your reviewer:

This (RFC) PR consists of two separate pieces (support RDT annotation(s) and integration with goresctrl) that could be split into separate PRs.

Related runc PR:
https://github.com/opencontainers/runc/pull/2920

Sibling PR against containerd:
https://github.com/containerd/containerd/pull/5439

/hold

#### Does this PR introduce a user-facing change?

```release-note
Support `io.kubernetes.cri.rdt-class` container annotation for specifying RDT class.

Add rdt_config_file config file option (and corresponding --rdt-config-file for command line) for configuring the resctrl pseudo-filesystem.
```

### CODE REVIEW

**Comment by haircommander** [comment]
cool! thanks for doing this work @marquiz. 

It is my understanding that this currently doesn't totally work (the rdt found in the annotation doesn't seem to be going anywhere). that's fine to start, just wanted to check.

Is it standard for all containers to have the same rdt profile for a host? or would they want the configuration to look differently? (I don't know anything about rdt)

If they may look different, it may behoove us to have mutliple profiles, found in a directory, rather than a single file. seccomp does this currently, where if you specify `localhost/filename` it looks for the seccomp config file in the filename in a specific directory.

I have a few review comments and nits, and we'll likely want some tests to go along with this one day. I would be interested in adding support independent of corresponding CRI changes. our `allowed_annotations` mechanism allows admins to specify annotations on the runtime class they want to allow. This means they can opt-out of this behavior, so we don't risk anything by adding it

**Comment by marquiz** [comment]
Thanks for taking a look at this @haircommander !

> It is my understanding that this currently doesn't totally work (the rdt found in the annotation doesn't seem to be going anywhere). that's fine to start, just wanted to check.

It does work. We inspect the annotation(s) and set `Spec.Linux.IntelRdt.ClosID` field accordingly. There is a small catch, though, as runc does not correctly support the `ClosID` field. That's what the related PR I noted (https://github.com/opencontainers/runc/pull/2920) fixes – I haven't taken a look at crun, yet.

> Is it standard for all containers to have the same rdt profile for a host? or would they want the configuration to look differently? (I don't know anything about rdt)
> 
> If they may look different, it may behoove us to have mutliple profiles, found in a directory, rather than a single file. seccomp does this currently, where if you specify `localhost/filename` it looks for the seccomp config file in the filename in a specific directory.

RDT enables QoS control of cache and/or memory bandwidth by providing class-based allocation of these resources. IOW, it allows a set of Classes-of-Service (CLOSes) that cache lines and/or memory bandwidth is allocated. PIDs are then assigned to one of those classes limiting their resource usage. The maximum number of classes is fairly limited (by HW) e.g. to 16.

The OCI container runtime (runc) does not read any configuration file. It just assigns the container (PIDs) to the class/clos specified in the container runtime spec. With this PR, classes/closes are pre-configured at CRI-O startup by goresctrl. Or probably better, "may be configured", as the class/clos configuration could be done out-of-band, manually or with some other tool as well.

I'm not sure of the aspect of independently configured profiles of e.g. seccomp/apparmor is applicable here. At least not in the case of goresctrl. The classes are inter-dependent as you may e.g. specify that reserve 50% to class A exclusively and remaining 50% to B and C so that C only gets 60% of that share. The set of available classes/closes are specified in a single configuration file.

Some references:
- https://01.org/intel-rdt-linux/blogs/fyu1/2017/resource-allocation-intel%C2%AE-resource-director-technology
- https://www.kernel.org/doc/html/latest/x86/resctrl.html
 
> I have a few review comments and nits, and we'll likely want some tests to go along with this one day.

I didn't add any, yet, as the integration is so "thin". We could basically test the config parsing error cases and parsing of the annotation(s) but that's about it.

> I would be interested in adding support independent of corresponding CRI changes. our `allowed_annotations` mechanism allows admins to specify annotations on the runtime class they want to allow. This means they can opt-out of this behavior, so we don't risk anything by adding it

Yeah, I added the container annotation to the list of `allowed_annotations`.

I think that the last patch (adding Pod annotations) really should not go into CRI-O. I'll submit that against K8s/kubelet if this approach looks feasible to the container runtimes (I submitted a PR against containerd as well: https://github.com/containerd/containerd/pull/5439)

**Comment by marquiz** [comment]
Rebased

File: pkg/config/config.go [hunk 1]
```diff
@@ -116,6 +117,11 @@ const (
 	DefaultIrqBalanceConfigFile = "/etc/sysconfig/irqbalance"
 )
 
+const (
+	// DefaultRdtConfigFile is the default value for RDT config file path
+	DefaultRdtConfigFile = ""
+)
+
```
Comment (saschagrunert) [inline pkg/config/config.go:124]: Let's move this const to `internal/config/rdt/rdt.go`.

File: pkg/config/template.go [hunk 2]
```diff
@@ -735,6 +740,11 @@ irqbalance_config_file = "{{ .IrqBalanceConfigFile }}"
 
 `
 
+const templateStringCrioRuntimeRdtConfigFile = `# Path to the RDT configuration file for configuring the resctrl pseudo-filesystem.
+rdt_config_file = "{{ .RdtConfigFile }}"
+
```
Comment (saschagrunert) [inline pkg/config/template.go:745]: Let's mention that this support configuration reload, too.

**Review by saschagrunert — COMMENTED** [review]
Looks good from my point of view. Can we test this somehow in our CI?

**Comment by TomSweeneyRedHat** [comment]
Other than @saschagrunert 's comments, LGTM

**Comment by saschagrunert** [comment]
@marquiz please rebase again :upside_down_face:

File: pkg/config/config.go (hunk 1 above)
Comment (marquiz) [reply pkg/config/config.go]: Moved

File: pkg/config/template.go (hunk 2 above)
Comment (marquiz) [reply pkg/config/template.go]: Addressed

**Comment by saschagrunert** [comment]
> The functionality is very much dependent on the HW feature. I'd say that in unit testing it is only possible/meaningful to test config parsing and parsing of the annotation(s). I could add those.

Sounds good to me, thank you! :pray: 

> Deeper integration tests would require running on bare-metal HW with RDT enabled.
> 
> WDYT?

Hm, yeah I assumed that it would require bare-metal testing. Should be fine for now as it is.

> What's your opinion on the last patch adding Pod annotations (that would/should be handled by kubelet). I'd prolly drop it from here

Yes I think we can drop the patch, WDYT @haircommander ?

**Comment by haircommander** [comment]
> > What's your opinion on the last patch adding Pod annotations (that would/should be handled by kubelet). I'd prolly drop it from here
> 
> Yes I think we can drop the patch, WDYT @haircommander ?

I kind of like having the patch there. What is the timeline for first-class support in the kubelet? to me, it's an idiomatic convention (appending /pod or /ctr-name) that provides a good user interface for little code addition

**Comment by kad** [comment]
@haircommander we wanted to start with blueprint of seccomp --> implement all needed plumbing in runtimes, add annotation first and after one/max two k8s release cycles target to podspec field to define both cache classes and in related PR #4873 blockio. Both of those should be similar in UI/implementation.

**Comment by haircommander** [comment]
> @haircommander we wanted to start with blueprint of seccomp --> implement all needed plumbing in runtimes, add annotation first and after one/max two k8s release cycles target to podspec field to define both cache classes and in related PR #4873 blockio. Both of those should be similar in UI/implementation.

cool, so I think keeping the structure of the annotations similar to seccomp makes sense, which indicates we should keep the last commit. wdyt @saschagrunert

**Comment by marquiz** [comment]
> cool, so I think keeping the structure of the annotations similar to seccomp makes sense, which indicates we should keep the last commit. wdyt @saschagrunert

👍 Let's keep it if you're cool with it. It can be removed after kubelet support is there.

**Comment by marquiz** [comment]
ping, any comments on the updates?

File: server/container_create_linux_test.go [hunk 3]
```diff
@@ -71,3 +72,35 @@ func TestAddOCIBindsForSys(t *testing.T) {
 		t.Error("there is not a single /sys bind mount")
 	}
 }
+
+func TestRdtClassFromAnnotations(t *testing.T) {
```
Comment (haircommander) [inline server/container_create_linux_test.go]: nit: we should be moving more tests to ginkgo, rather than testing. can these be written using ginkgo

**Comment by haircommander** [comment]
+1 to moving the functionality to the shared library!
I left one nit, but it generally looks good (once the commits are squashed a bit)
the docs and completions aren't correct you'll want to run:
`make docs-generation && make completions-generation` followed by `make docs-validation && make completions-validation`
(note these make targets take the host crio config into account, so there may be unintended changes that you'll need to drop)

File: server/container_create_linux_test.go (hunk 3 above)
Comment (marquiz) [reply server/container_create_linux_test.go]: These tests were dropped (moved over to goresctrl codebase) when moving the annotation parsing logic there

**Comment by marquiz** [comment]
> I left one nit, but it generally looks good (once the commits are squashed a bit)

Thanks for the review. I now rebased and squashed this into a more meaningful patchset. I still kept the resctrl configuration and annotation parsing as separate patches.

> the docs and completions aren't correct you'll want to run:
> `make docs-generation && make completions-generation` followed by `make docs-validation && make completions-validation`
> (note these make targets take the host crio config into account, so there may be unintended changes that you'll need to drop)

Thanks for the note. I did as instructed (except `completions-validation` make target does not seems to exist) and those should be in line, now.

File: internal/config/rdt/rdt.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+	logrus.Infof("RDT enabled, config successfully loaded from %q", path)
+	c.enabled = true
+	c.config = tmpCfg
+
+	return nil
+}
+
+func loadConfigFile(path string) (*rdt.Config, error) {
+	data, err := ioutil.ReadFile(path)
```
[the remark this answers is not in the document]
Comment (marquiz) [reply internal/config/rdt/rdt.go:83]: I'd ignore this as the config from sysadmin is supposed to be trusted input

**Comment by marquiz** [comment]
Rebased.

> The first one is introduciing separate "supported" (by system) and "enabled" (in cri-o) status for RDT. Related to this, we now simply ignore the class (set via annotation) if RDT is disabled. Should we return an error instead? 🤔

Answered to myself: let's return an error. Changed this so that we now return an error if RDT class would be set via an annotation but RDT has been disabled.

**Comment by kolyshkin** [comment]
@marquiz do you think you can fix the issues with goresctrl (see comments from sonatype-lift bot above) and tag a release in there, so this PR can use a tagged version?

**Comment by marquiz** [comment]
> @marquiz do you think you can fix the issues with goresctrl (see comments from sonatype-lift bot above) and tag a release in there, so this PR can use a tagged version?

I believe so 😊  I'll update deps in goresctrl, update this PR (with that version) and let's see what sonatype says. If she's happy I'll tag goresctrl and update this PR one more time with the tagged version

**Comment by marquiz** [comment]
Hmm, seems like sonatype is not happy 😞  I think managing the indirect dependencies manually in goresctrl would be unmaintainable and error prone (e.g. go mod tidy would clean those away). Also, I checked that the listed deps above are already (indirect) dependencies of cri-o at present. I'd suggest to ignore the comments (especially as cri-o already depends on those vulnerabilities). WDYT @kolyshkin?

**Comment by askervin** [comment]
@kolyshkin, @marquiz, @kad, I inspected dependencies that bring in the three `x/net` module versions with 5 critical vulnerabilities. (I'm using a [go-mod-tree](https://github.com/intel/cri-resource-manager/pull/664/files) script to help viewing the dependency graph.)

Here are shortest paths from CRI-O to each of the vulnerable x/net module versions (golang.org/x/net@0.0.0...-8a410e7b638d, ...161cd47e91fd and ...3673e40ba225)
```
$ go mod graph | go-mod-tree --shortest-path '.*(8a410e7b638d|161cd47e91fd|3673e40ba225)'
L1D0    github.com/cri-o/cri-o
L2D1    :   github.com/containerd/ttrpc@v1.0.2
L3D2    :   :   google.golang.org/genproto@v0.0.0-20200117163144-32f20d992d24
L4D3    :   :   :   google.golang.org/grpc@v1.19.0
L5D4    :   :   :   :   golang.org/x/net@v0.0.0-20180826012351-8a410e7b638d
L6D1    :   github.com/cri-o/ocicni@v0.2.1-0.20210301205850-541cf7c703cf
L7D2    :   :   github.com/onsi/gomega@v1.5.0
L8D3    :   :   :   golang.org/x/net@v0.0.0-20180906233101-161cd47e91fd
L9D1    :   k8s.io/client-go@v0.21.1
L10D2   :   :   golang.org/x/oauth2@v0.0.0-20200107190931-bf48bf16ab8d
L11D3   :   :   :   google.golang.org/appengine@v1.4.0
L12D4   :   :   :   :   golang.org/x/net@v0.0.0-20180724234803-3673e40ba225
```

Here are *all* CRI-O's direct dependencies that later on lead to importing these x/net versions:
```
$ go mod graph | go-mod-tree  --to '.*(8a410e7b638d|161cd47e91fd|3673e40ba225)' --depth 1
L1D0    github.com/cri-o/cri-o
L2D1    :   github.com/containerd/containerd@v1.5.2
L3D1    :   github.com/containerd/ttrpc@v1.0.2
L4D1    :   github.com/containernetworking/plugins@v0.9.1
L5D1    :   github.com/containers/buildah@v1.20.1
L6D1    :   github.com/containers/common@v0.37.1
L7D1    :   github.com/containers/image/v5@v5.13.2
L8D1    :   github.com/containers/ocicrypt@v1.1.1
L9D1    :   github.com/containers/podman/v3@v3.1.2
L10D1   :   github.com/containers/storage@v1.32.2
L11D1   :   github.com/cri-o/ocicni@v0.2.1-0.20210301205850-541cf7c703cf
L12D1   :   github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0
L13D1   :   github.com/intel/goresctrl@v0.0.0-20210623080121-be5fa857f4b4
L14D1   :   github.com/onsi/ginkgo@v1.16.4
L15D1   :   github.com/onsi/gomega@v1.13.0
L16D1   :   github.com/opencontainers/runc@v1.0.0-rc95.0.20210521141834-a95237f81684
L17D1   :   github.com/prometheus/client_golang@v1.11.0
L18D1   :   github.com/prometheus/common@v0.29.0
L19D1   :   google.golang.org/grpc@v1.38.0
L20D1   :   k8s.io/api@v0.21.1
L21D1   :   k8s.io/apimachinery@v0.21.2
L22D1   :   k8s.io/client-go@v0.21.1
L23D1   :   k8s.io/cri-api@v0.21.1
L24D1   :   k8s.io/kubernetes@v1.21.1
L25D1   :   k8s.io/release@v0.8.0
L26D1   :   sigs.k8s.io/release-utils@v0.2.1
L27D1   :   sigs.k8s.io/zeitgeist@v0.3.0
```

And the other way around, these are the modules that directly import the vulnerable x/net versions:
```
$ go mod graph | go-mod-tree --reverse --from '.*(8a410e7b638d|161cd47e91fd|3673e40ba225)' --depth 1
L1D0    golang.org/x/net@v0.0.0-20180724234803-3673e40ba225
L2D1    :   google.golang.org/appengine@v1.3.0
L3D1    :   google.golang.org/appengine@v1.4.0
L4D1    :   google.golang.org/appengine@v1.5.0
L5D0    golang.org/x/net@v0.0.0-20180826012351-8a410e7b638d
L6D1    :   google.golang.org/grpc@v1.19.0
L7D0    golang.org/x/net@v0.0.0-20180906233101-161cd47e91fd
L8D1    :   github.com/containerd/continuity@v0.0.0-20201208142359-180525291bb7
L9D1    :   github.com/containerd/continuity@v0.0.0-20210208174643-50096c924a4e
L10D1   :   github.com/lucas-clemente/quic-go@v0.10.2
L11D1   :   github.com/onsi/gomega@v1.5.0
L12D1   :   github.com/onsi/gomega@v1.7.0
L13D1   :   github.com/onsi/gomega@v1.7.1
L14D1   :   github.com/onsi/gomega@v1.8.1
L15D1   :   github.com/onsi/gomega@v1.9.0
```

I could not get rid of the x/net versions with
```
replace (
        golang.org/x/net => golang.org/x/net v0.0.0-20210525063256-abc453219eb5
```

However, replacing (updating) libraries that directly imported bad x/net versions seems to solve the problem. That is, this works:
```
replace (
        github.com/containerd/continuity => github.com/containerd/continuity v0.1.0
        github.com/lucas-clemente/quic-go => github.com/lucas-clemente/quic-go v0.21.1
        github.com/onsi/gomega => github.com/onsi/gomega v1.13.0
        go.opencensus.io => go.opencensus.io v0.23.0
        google.golang.org/appengine => google.golang.org/appengine v1.6.7
        google.golang.org/grpc => google.golang.org/grpc v1.39.0
```

What do you think, should we simply update these modules? Or do more thorough dependency scrubbing - including modules that import these modules, etc.?

**Comment by marquiz** [comment]
> However, replacing (updating) libraries that directly imported bad x/net versions seems to solve the problem. That is, this works:
> 
> ```
> replace (
>         github.com/containerd/continuity => github.com/containerd/continuity v0.1.0
>         github.com/lucas-clemente/quic-go => github.com/lucas-clemente/quic-go v0.21.1
>         github.com/onsi/gomega => github.com/onsi/gomega v1.13.0
>         go.opencensus.io => go.opencensus.io v0.23.0
>         google.golang.org/appengine => google.golang.org/appengine v1.6.7
>         google.golang.org/grpc => google.golang.org/grpc v1.39.0
> ```
> 
> What do you think, should we simply update these modules? Or do more thorough dependency scrubbing - including modules that import these modules, etc.?

Thanks @askervin for digging into this ⛏️  👍  

I really have no good idea how this should be resolved, other than submitting PRs against all the problematic dependencies we have. The "manual replace method" looks unmaintainable and error prone in the long run 🤔 

In any case this should be a subject of a separate PR, IMO. Any comments @kolyshkin

**Comment by marquiz** [comment]
Rebased

**Comment by marquiz** [comment]
Rebased

> Hmm, seems like sonatype is not happy 😞 I think managing the indirect dependencies manually in goresctrl would be unmaintainable and error prone (e.g. go mod tidy would clean those away). Also, I checked that the listed deps above are already (indirect) dependencies of cri-o at present. I'd suggest to ignore the comments (especially as cri-o already depends on those vulnerabilities).

Any thoughts/advice on this, anyone? @kolyshkin @haircommander @saschagrunert ?

**Comment by saschagrunert** [comment]
> Rebased

I'm sorry we have to rebase again after the k8s bump.

> > Hmm, seems like sonatype is not happy. I think managing the indirect dependencies manually in goresctrl would be unmaintainable and error prone (e.g. go mod tidy would clean those away). Also, I checked that the listed deps above are already (indirect) dependencies of cri-o at present. I'd suggest to ignore the comments (especially as cri-o already depends on those vulnerabilities).

Yes, I agree.

**Comment by marquiz** [comment]
> I'm sorry we have to rebase again after the k8s bump.

Np, this was an easy one.

> > Hmm, seems like sonatype is not happy. I think managing the indirect dependencies manually in goresctrl would be unmaintainable and error prone (e.g. go mod tidy would clean those away). Also, I checked that the listed deps above are already (indirect) dependencies of cri-o at present. I'd suggest to ignore the comments (especially as cri-o already depends on those vulnerabilities).
> 
> Yes, I agree.

👍

**Comment by haircommander** [comment]
/assign @saschagrunert 
for the final approval

**Review by saschagrunert — APPROVED** [review]
/lgtm
/unhold

**Comment by haircommander** [comment]
/override ci/prow/e2e-gcp

agnostic passed

**Pull request merged**